### DISCUSSION TITLE
How to handle submit errors (not validation)?

### DISCUSSION THREAD

**Discussion by wKovacs64** [react-hook-form/react-hook-form#5803] [Q&A] [answered]
Hi. I'm wondering how to handle errors thrown in the submit handler passed to `handleSubmit()`. I expected RHF to catch errors thrown in that function for me (or something similar) and set `isSubmitSuccessful` to `false`, but it doesn't appear to work that way. All the documentation I read seemed to speak to validation or individual inputs, but I'm just trying to handle the scenario where the submit fails for some reason (e.g. fetch call throws).

Here's an example to demonstrate:

https://codesandbox.io/s/react-hook-form-basic-forked-9i3f7?file=/src/index.js

If you submit the form, the submit handler throws, and it is considered an unhandled exception (you can tell because Create React App pops up its error overlay in development, and in production it is logged to the console as an uncaught error). If I wrap the error in a try-catch block to prevent this, RHF doesn't know an error occurred (of course) so it thinks the submit was successful. The only manual way I saw to tell RHF that something failed was `setError()`, but this seems to pertain to a particular input (the first parameter is an input name).

I'm not trying to handle validation or individual inputs, just failed submissions. What is the recommended pattern here?

Thanks.

**Comment by bluebill1049** [comment, marked as answer]
is this what you aftrer? https://react-hook-form.com/api/useform/seterror

**Reply by wKovacs64** [reply]
I don't think so? That looks like it's for setting an error for a particular input and all my inputs are fine. Maybe I'm misunderstanding the usage, though?

**Reply by bluebill1049** [reply]
i see, we don't capture the error instead let the error throw and be caught, and also mark the submitted as failed.

https://github.com/react-hook-form/react-hook-form/blob/master/src/useForm.ts#L1032

**Reply by wKovacs64** [reply]
OK, but then I need to catch the error - where would you recommend I do that? If I catch it inside the submit handler, RHF won't know it threw so it won't set `isSubmitSuccessful` to `false`. And if I don't catch it, it's an uncaught exception.

I could wrap `handleSubmit()` but that feels extraneous and I haven't seen that in any of the examples. I could also use an Error Boundary, but I'd rather be able to handle it inside the form. 🤔 

Edit: those might not even be solutions, actually.

**Reply by bluebill1049** [reply]
here is a solution: https://codesandbox.io/s/react-hook-form-basic-forked-mspf5?file=/src/index.js

```tsx
import React from "react";
import ReactDOM from "react-dom";
import { useForm } from "./src";

import "./index.css";

function App() {
  const {
    register,
    handleSubmit,
    formState: { isSubmitSuccessful }
  } = useForm();

  const onSubmit = async () => {
    throw new Error("test");
  };

  console.log("isSubmitSuccessful", isSubmitSuccessful);

  return (
    <div className="App">
      <form
        onSubmit={(e) => {
          handleSubmit(onSubmit)(e).catch((e) => {
            console.log("hi", e);
          });
        }}
      >
        <div>
          <label htmlFor="firstName">First Name</label>
          <input placeholder="bill" {...register("firstName")} />
        </div>

        <input type="submit" />
      </form>
    </div>
  );
}

const rootElement = document.getElementById("root");
ReactDOM.render(<App />, rootElement);
```

i don't think hook form would swallow the error which threw by the user's code.

**Reply by wKovacs64** [reply]
Hmm, ok, thanks - I'll try that tomorrow. Is this the approach you would recommend to handle submission errors?

**Reply by wKovacs64** [reply]
Confirmed this is working, by the way. Thanks.

**Reply by bluebill1049** [reply]
no worries 👍

**Reply by Noitidart** [reply]
How can I throw field errors from this? Like if multiple fields from server is bad, server side should say so.

I was trying this but I keep getting `path.split` is not a function:

```
form.setError([
{ type: 'server', name: 'color', message: 'foo' },
{ type: 'server', name: 'height', message: 'bar' }
])
```

I got this code from the form.setError example here:

<img width="1311" alt="image" src="https://user-images.githubusercontent.com/6372489/129135912-f2122c1d-88c5-48c7-b6b8-0a9e7377d50d.png">
